idx/bigtable/bigtable.go

```diff
@@ -0,0 +1,549 @@
+		rowKeys := make([]string, len(buffer))
+		mutations := make([]*bigtable.Mutation, len(buffer))
+		byteCount := 0
+		for i, req := range buffer {
+			statQueryInsertWaitDuration.Value(time.Since(req.recvTime))
+			key, cols := SchemaToRow(req.def)
+			rowKeys[i] = key
+			mut := bigtable.NewMutation()
+			for col, val := range cols {
+				mut.Set(COLUMN_FAMILY, col, bigtable.Now(), val)
+				byteCount += len(val)
+			}
+			mutations[i] = mut
+		}
+		statSaveBytesPerRequest.Value(byteCount)
+		pre := time.Now()
+		success := false
+		attempts := 0
+
+		for !success {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+			errs, err := b.tbl.ApplyBulk(ctx, rowKeys, mutations)
+			cancel()
+			if err != nil {
+				log.Errorf("bigtable-idx: unable to apply writes to bigtable. %s", err)
+				statQueryInsertFail.Add(len(rowKeys))
+				if (attempts % 20) == 0 {
+					log.Warnf("bigtable-idx: Failed to write %d defs to bigtable. they will be retried. %s", len(rowKeys), err)
+				}
+				sleepTime := 100 * attempts
+				if sleepTime > 2000 {
+					sleepTime = 2000
+				}
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+				attempts++
+			} else if len(errs) > 0 {
+				failedRowKeys := make([]string, 0)
+				failedMutations := make([]*bigtable.Mutation, 0)
+				for i, err := range errs {
+					if err != nil {
+						failedRowKeys = append(failedRowKeys, rowKeys[i])
+						failedMutations = append(failedMutations, mutations[i])
+					}
+				}
+				rowKeys = failedRowKeys
+				mutations = failedMutations
+				log.Errorf("bigtable-idx: failed to write %d rows. %s", len(failedRowKeys), err)
+				statQueryInsertFail.Add(len(failedRowKeys))
+				sleepTime := 100 * attempts
+				if sleepTime > 2000 {
+					sleepTime = 2000
+				}
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+				attempts++
+			} else {
+				success = true
+				statQueryInsertExecDuration.Value(time.Since(pre))
+				statQueryInsertOk.Add(len(rowKeys))
+				log.Debugf("bigtable-idx: %d metricDefs saved to bigtable.", len(rowKeys))
+			}
+
+		}
+		buffer = buffer[:0]
+	}
+
+LOOP:
+	for {
+		select {
+		case <-timer.C:
+			timer.Reset(time.Second)
+			if len(buffer) > 0 {
+				flush()
+			}
+		case req, ok := <-b.writeQueue:
+			if !ok {
+				break LOOP
+			}
+			buffer = append(buffer, req)
+			if len(buffer) >= writeMaxFlushSize {
+				// make sure the timer hasn't already fired. If it has we read
+				// from the chan and consume the event.
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(time.Second)
+				flush()
+			}
+		}
+	}
+	log.Info("bigtable-idx: writeQueue handler ended.")
+	b.wg.Done()
+}
+
+func (b *BigtableIdx) Delete(orgId uint32, pattern string) ([]idx.Archive, error) {
+	pre := time.Now()
+	defs, err := b.MemoryIdx.Delete(orgId, pattern)
+	if err != nil {
+		return defs, err
+	}
+	if updateIdx {
+		for _, def := range defs {
+			err = b.deleteDef(&def.MetricDefinition)
+			if err != nil {
+				break
+			}
```

Comment:
the equivalent code in cass idx is:
```
                for _, def := range defs {
                        err = c.deleteDef(def.Id, def.Partition)
                        if err != nil {
                                log.Errorf("cassandra-idx: %s", err.Error())
                        }
                }
```
in callers we don't tend to log these errors, so we should log it here

Reply:
this has been addressed.